### PR TITLE
Feature/paragraph component

### LINKED ISSUE #223: Discussion: how to handle a paragraph/text component?
How would we support a parent application being able to form a block of text that supports `a`, `em` child elements?

Issues to consider:

- If we allow markdown to be passed in as a prop then we'd need to bundle a markdown parser, which would be a large amount of JS
- If we rely on react-dom elements it may not be portable to react-native
- Would a Text/Paragraph component allow other React components as children or should it be a leaf node/escape hatch? If we style HTML elements, they will cascade if the Text or Paragraph component has other React component children. If we don't allow child elements, how do we handle the need for a React Router link?

### PR DESCRIPTION
**Pull request by Loque-**
Adds paragraph component with react-markdown, restricted to only allow Links, Bold and Italics. Whilst GDS advises against Bold and Italics they should still be supported to ensure the component can be used by as many teams as possible.

This PR also contains ESLint fixes (apologies for mixing them into this PR)

closes #223 

* [x] Documentation
* [x] Tests
* [x] Ready to be merged

### CODE REVIEW

File: components/paragraph/src/link-renderer.js [hunk 1]
```diff
@@ -0,0 +1,26 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import { Link } from 'react-router-dom';
+import Anchor from './atoms/anchor';
+
+// Support react-router links
+// https://github.com/rexxars/react-markdown/issues/29#issuecomment-231556543
+function LinkRenderer({ href, children }) {
```
Comment (stevesims) [inline components/paragraph/src/link-renderer.js]: as a functional component it's probably more conventional for this to be an arrow function :grin:

File: components/paragraph/src/index.js [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+   */
+  supportingText: PropTypes.bool,
+};
+
+Paragraph.defaultProps = {
+  children: '',
+  supportingText: false,
+};
+
+export default withWhiteSpace({ marginBottom: 5 })(Paragraph);
```
Comment (stevesims) [inline components/paragraph/src/index.js]: isn't the syntax for `withWhiteSpace` to use `mb: 5`?

File: components/paragraph/src/link-renderer.js [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+  return (
+    href.match(/^\//)
+      ? <Link to={href}>{children}</Link>
+      : <Anchor href={href}>{children}</Anchor>
+  );
+}
+
+LinkRenderer.propTypes = {
+  href: PropTypes.string,
+  children: PropTypes.string,
```
Comment (stevesims) [inline components/paragraph/src/link-renderer.js]: `children` should be `PropTypes.node`

File: components/paragraph/src/test.js [hunk 4]
```diff
@@ -0,0 +1,47 @@
+import React from 'react';
+import ReactDOM from 'react-dom';
+import { MemoryRouter } from 'react-router-dom';
+import { mount } from 'enzyme';
+
+import Paragraph from './';
+import LinkRenderer from './link-renderer';
```
Comment (stevesims) [inline components/paragraph/src/test.js]: we should consider splitting the `LinkRenderer` tests out to a separate test file

File: components/paragraph/src/test.js [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+    ReactDOM.render(<Paragraph>Lorem ipsum</Paragraph>, div);
+  });
+
+  it('matches wrapper snapshot', () => {
+    expect(mount(wrapper)).toMatchSnapshot('wrapper mount');
+  });
+});
+
+describe(LinkRenderer, () => {
+  it('returns a react-router `Link` with a relative href path', () => {
```
Comment (stevesims) [inline components/paragraph/src/test.js]: consider adding a test for `#anchor` links

File: components/paragraph/src/index.js (hunk 2 above)
Comment (Loque-) [reply components/paragraph/src/index.js]: `mb=""` is used as an attribute on JSX components, but when added as a HoC the attribute name appears to be `marginBottom`; https://github.com/penx/govuk-react/blob/master/components/button/src/index.js#L86

File: components/paragraph/src/link-renderer.js (hunk 3 above)
Comment (Loque-) [reply components/paragraph/src/link-renderer.js]: I would rather it stayed as a string as I only want the component to support strings and not nested nodes. AFAIK Markdown should be treated as a string.

File: components/paragraph/src/test.js (hunk 4 above)
Comment (Loque-) [reply components/paragraph/src/test.js]: Definitely, thanks for this and the ticket!

File: components/paragraph/src/link-renderer.js (hunk 1 above)
Comment (Loque-) [reply components/paragraph/src/link-renderer.js]: Great shout, thank you!

File: components/paragraph/src/test.js (hunk 5 above)
Comment (Loque-) [reply components/paragraph/src/test.js]: Great shout, thank you, added!

File: components/paragraph/src/index.js (hunk 2 above)
Comment (penx) [reply components/paragraph/src/index.js]: Perhaps we need an issue to review consistency in future

File: components/paragraph/package.json [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+  "name": "@govuk-react/paragraph",
+  "version": "0.2.1",
+  "private": false,
+  "publishConfig": {
+    "access": "public"
+  },
+  "dependencies": {
+    "@govuk-react/constants": "^0.2.1",
+    "@govuk-react/hoc": "^0.2.1",
+    "@storybook/addon-knobs": "^3.4.2",
```
Comment (penx) [inline components/paragraph/package.json]: Storybook shouldn’t be a production dependency

File: components/paragraph/src/index.js (hunk 2 above)
Comment (stevesims) [reply components/paragraph/src/index.js]: yeah maybe - the difference is a little bit confusing :grin:

but i am often easily confused... 😕

File: components/paragraph/package.json (hunk 6 above)
Comment (Loque-) [reply components/paragraph/package.json]: I noticed this in some other files, I'll make a note to move it from these to devDependancies also. Cheers for the review!

**Pull request merged**